File: sdk/py/tutorials/vision/tutorialVisionDataAugmentation.py
#!/usr/bin/env python3
"""Render six deterministic GPU augmentation views from one source image."""

from __future__ import annotations

import argparse

# pyright: reportWildcardImportFromLibrary=false
from oa import *


def main() -> None:
	parser = argparse.ArgumentParser()
	parser.add_argument(
		"--input",
		default=Paths.asset("image/visionTestPattern320x180.jpg"),
	)
	parser.add_argument(
		"--output",
		default="/tmp/oa-vision-data-augmentation.png",
	)
	args = parser.parse_args()

	if not initComputeEngine():
		raise RuntimeError("OA could not create a Vulkan compute engine")

	try:
		sourceImage = FnImage.decodeFile(args.input)
		source = sourceImage.asMatrix()

		# OA_DOC_BEGIN: vision-data-augmentation
		views = [sourceImage]
		views.append(Image(
			FnImage.flip(source, True, False),
			ImageLayout.Nchw,
			ImageFormat.Rgb,
		))
		crop = FnImage.centerCrop(source, 280, 150)
		views.append(Image(
			FnImage.resize(crop, 320, 180),
			ImageLayout.Nchw,
			ImageFormat.Rgb,
		))
		views.append(Image(
			FnImage.brightnessContrast(source, 0.08, 1.15),
			ImageLayout.Nchw,
			ImageFormat.Rgb,
		))
		noisy = FnImage.gaussianNoise(source, 0.0, 0.035, 2026)
		views.append(Image(
			FnImage.clamp(noisy, 0.0, 1.0),
			ImageLayout.Nchw,
			ImageFormat.Rgb,
		))
		views.append(Image(
			FnImage.solarize(source, 0.55, 1.0),
			ImageLayout.Nchw,
			ImageFormat.Rgb,
		))
		# OA_DOC_END: vision-data-augmentation

		titles = [
			"source",
			"horizontal flip",
			"crop + resize",
			"brightness + contrast",
			"seeded Gaussian noise",
			"solarize",
		]
		config = plot.FigureConfig()
		config.title = "OA GPU data augmentation"
		config.rows = 2
		config.cols = 3
		config.width = 1200
		config.height = 760
		config.hSpacing = 18
		config.vSpacing = 20
		config.padding = 20
		config.theme = plot.Theme.Dark
		figure = plot.Figure(config)
		for index, (title, image) in enumerate(zip(titles, views, strict=True)):
			axes = figure.ax(index // 3, index % 3)
			axes.imshow(image)
			axes.title(title)
			axes.grid(False)
			axes.legend(False)

		figure.saveTo(args.output)
		print(f"saved 6 GPU augmentation views to {args.output}")
	finally:
		shutdownComputeEngine()


if __name__ == "__main__":
	main()

File: sdk/cpp/tutorials/vision/tutorialVisionDataAugmentation.cpp
// OA Vision tutorial — deterministic GPU data-augmentation views.

#include <oa/core/filesystem.h>
#include <oa/core/paths.h>
#include <oa/runtime/engine.h>
#include <oa/runtime/texture.h>
#include <oa/ui/plot/plot.h>
#include <oa/vision/fnImage.h>

namespace {

oa::Image rgbImage(oa::Matrix inMatrix) {
	return oa::Image(
		oa::move(inMatrix),
		oa::ImageLayout::Nchw,
		oa::ImageFormat::Rgb);
}

} // namespace

int main(int argc, char** argv) {
	const oa::Path input = argc > 1
		? oa::Path(argv[1])
		: oa::Paths::asset("image/visionTestPattern320x180.jpg");
	const oa::Path output = argc > 2
		? oa::Path(argv[2])
		: oa::Path("/tmp/oa-vision-data-augmentation.png");

	oa::EngineConfig config;
	config.appName = "TutorialVisionDataAugmentation";
	config.presentationMode = oa::PresentationMode::Headless;
	auto created = oa::Engine::create(config);
	if (not created.isOk()) {
		oa::print(oa::PrintStream::Error, "Engine creation failed: {}",
			created.getStatus().toString().cStr());
		return 1;
	}
	auto engine = oa::move(created).getValue();

	auto decoded = oa::FnImage::decodeFile(input);
	if (not decoded.isOk()) {
		oa::print(oa::PrintStream::Error, "Image decode failed: {}",
			decoded.getStatus().toString().cStr());
		return 1;
	}

	// OA_DOC_BEGIN: vision-data-augmentation
	oa::Vector<oa::Image> views;
	views.reserve(6);
	views.pushBack(oa::move(decoded).getValue());
	const oa::Matrix& source = views.front().asMatrix();

	views.pushBack(rgbImage(oa::FnImage::flip(source, true, false)));
	auto crop = oa::FnImage::centerCrop(source, 280, 150);
	views.pushBack(rgbImage(oa::FnImage::resize(crop, 320, 180)));
	views.pushBack(rgbImage(
		oa::FnImage::brightnessContrast(source, 0.08F, 1.15F)));
	auto noisy = oa::FnImage::gaussianNoise(source, 0.0F, 0.035F, 2026U);
	views.pushBack(rgbImage(oa::FnImage::clamp(noisy, 0.0F, 1.0F)));
	views.pushBack(rgbImage(oa::FnImage::solarize(source, 0.55F, 1.0F)));
	// OA_DOC_END: vision-data-augmentation

	const char* titles[] = {
		"source", "horizontal flip", "crop + resize",
		"brightness + contrast", "seeded Gaussian noise", "solarize",
	};
	oa::plot::Figure figure({
		.title = "OA GPU data augmentation",
		.rows = 2,
		.cols = 3,
		.width = 1200,
		.height = 760,
		.hSpacing = 18,
		.vSpacing = 20,
		.padding = 20,
		.theme = oa::plot::Theme::Dark,
	});

	oa::Vector<oa::Texture> textures;
	textures.reserve(views.size());
	for (oa::Usize index = 0; index < views.size(); ++index) {
		auto texture = oa::FnTexture::fromImage(*engine, views[index]);
		if (not texture.isOk()) {
			oa::print(oa::PrintStream::Error, "texture conversion failed: {}",
				texture.getStatus().toString().cStr());
			return 1;
		}
		textures.pushBack(oa::move(texture).getValue());
		auto& axes = figure.ax(
			static_cast<oa::I32>(index / 3),
			static_cast<oa::I32>(index % 3));
		axes.imshow(textures.back());
		axes.title(titles[index]);
		axes.grid(false);
		axes.legend(false);
	}

	const oa::Path outputDirectory = output.parentPath();
	if (not outputDirectory.empty()) {
		if (auto status = oa::Filesystem::createDirectories(outputDirectory);
			not status.isOk()) {
			oa::print(oa::PrintStream::Error, "output directory creation failed: {}",
				status.toString().cStr());
			return 1;
		}
	}
	if (auto status = figure.saveTo(*engine, output.cStr()); not status.isOk()) {
		oa::print(oa::PrintStream::Error, "Figure save failed: {}",
			status.toString().cStr());
		return 1;
	}

	oa::print("saved 6 GPU augmentation views to {}", output.cStr());
	return 0;
}
